### ISSUE TITLE
Crash on iOS when using multiple engines spawned from FlutterEngineGroup

### ISSUE THREAD

**Issue by sroddy** [flutter/flutter#79335]
## Steps to Reproduce

1. Use the multiple_flutter sample
2. Update the main.dart file, by adding this code inside the _MyHomePageState initState:
```
    Timer.periodic(const Duration(milliseconds: 50), (timer) {
      _incrementCounter();
    });
``` 
3. Run in release mode
4. tap next until you see the double flutter view controllers (3 times)
5. go back and forth a couple of times.
6. the app will crash

<details>
  <summary>Logs</summary>

```
multiple-flutters.1.1.ui (9)#0	0x0000000105c48e88 in dart::Thread::top_exit_frame_info() const [inlined] at /opt/s/w/ir/cache/builder/src/third_party/dart/runtime/vm/thread.h:522
#1	0x0000000105c48e88 in dart::IsolateGroup::IncreaseMutatorCount(dart::Isolate*) [inlined] at /opt/s/w/ir/cache/builder/src/third_party/dart/runtime/vm/isolate.cc:683
#2	0x0000000105c48e80 in dart::Isolate::ScheduleThread(bool, bool) at /opt/s/w/ir/cache/builder/src/third_party/dart/runtime/vm/isolate.cc:3594
#3	0x0000000105d09d1c in dart::Thread::EnterIsolate(dart::Isolate*) [inlined] at /opt/s/w/ir/cache/builder/src/third_party/dart/runtime/vm/thread.cc:277
#4	0x0000000105d09d0c in ::Dart_EnterIsolate(Dart_Isolate) at /opt/s/w/ir/cache/builder/src/third_party/dart/runtime/vm/dart_api_impl.cc:1531
#5	0x0000000105a797fc in tonic::DartIsolateScope::DartIsolateScope(_Dart_Isolate*) [inlined] at /opt/s/w/ir/cache/builder/src/flutter/third_party/tonic/scopes/dart_isolate_scope.cc:16
#6	0x0000000105a797b4 in tonic::DartIsolateScope::DartIsolateScope(_Dart_Isolate*) at /opt/s/w/ir/cache/builder/src/flutter/third_party/tonic/scopes/dart_isolate_scope.cc:9
#7	0x0000000105a76f6c in tonic::DartPersistentValue::Clear() at /opt/s/w/ir/cache/builder/src/flutter/third_party/tonic/dart_persistent_value.cc:49
#8	0x0000000105a01d94 in std::__1::__function::__value_func<void ()>::operator()() const [inlined] at /opt/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/functional:1860
#9	0x0000000105a01d88 in std::__1::function<void ()>::operator()() const [inlined] at /opt/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/functional:2419
#10	0x0000000105a01d88 in fml::MessageLoopImpl::FlushTasks(fml::FlushType) at /opt/s/w/ir/cache/builder/src/flutter/fml/message_loop_impl.cc:130
#11	0x0000000105a03bf4 in fml::MessageLoopImpl::RunExpiredTasksNow() [inlined] at /opt/s/w/ir/cache/builder/src/flutter/fml/message_loop_impl.cc:143
#12	0x0000000105a03bec in fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) at /opt/s/w/ir/cache/builder/src/flutter/fml/platform/darwin/message_loop_darwin.mm:75
#13	0x000000018044dfa0 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ ()
#14	0x000000018044dba0 in __CFRunLoopDoTimer ()
#15	0x000000018044cffc in __CFRunLoopDoTimers ()
#16	0x0000000180446ee4 in __CFRunLoopRun ()
#17	0x000000018044621c in CFRunLoopRunSpecific ()
#18	0x0000000105a03ad0 in fml::MessageLoopDarwin::Run() at /opt/s/w/ir/cache/builder/src/flutter/fml/platform/darwin/message_loop_darwin.mm:46
#19	0x0000000105a034d4 in fml::MessageLoopImpl::DoRun() [inlined] at /opt/s/w/ir/cache/builder/src/flutter/fml/message_loop_impl.cc:96
#20	0x0000000105a034b8 in fml::MessageLoop::Run() [inlined] at /opt/s/w/ir/cache/builder/src/flutter/fml/message_loop.cc:49
#21	0x0000000105a034b4 in fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0::operator()() const [inlined] at /opt/s/w/ir/cache/builder/src/flutter/fml/thread.cc:35
#22	0x0000000105a03444 in decltype(std::__1::forward<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fp)()) std::__1::__invoke<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0&&) [inlined] at /opt/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:4425
#23	0x0000000105a03444 in void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>&, std::__1::__tuple_indices<>) [inlined] at /opt/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/thread:341
#24	0x0000000105a03444 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0> >(void*) at /opt/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/thread:351
#25	0x00000001cbf7fcb0 in _pthread_start ()
```

Tested on current stable and current dev builds

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel unknown, 2.0.3, on macOS 11.2.3 20D91 darwin-arm, locale it)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] IntelliJ IDEA Community Edition (version 2020.3.3)
[✓] VS Code (version 1.54.3)
[✓] Connected device (3 available)

• No issues found!
```

</details>

**Comment by sroddy** [comment]
cc @gaaclarke @xster

**Comment by xster** [comment]
@flutter-symbolizer-bot

**Comment by xster** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/79335#issue-844023356

**Comment by sroddy** [comment]
@xster here is a complete crash log taken from the device

[MultipleFluttersIos  30-03-21, 05-08.crash.zip](https://github.com/flutter/flutter/files/6228137/MultipleFluttersIos.30-03-21.05-08.crash.zip)

**Comment by TahaTesser** [comment]
Hi @sroddy 
Thanks for filing the issue, i can reproduce the crash on XR in profile mode from Xcode

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 2.0.3, on macOS 11.2.3 20D91 darwin-x64, locale
    en-GB)
    • Flutter version 2.0.3 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 4d7946a68d (11 days ago), 2021-03-18 17:24:33 -0700
    • Engine revision 3459eb2436
    • Dart version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.54.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (4 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E            • ios
      • iOS 14.4.1
    • iPhone 12 (mobile)     • 83060656-28E7-44CD-801E-B11EC3EC89BA • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-4 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64
      • macOS 11.2.3 20D91 darwin-x64
    • Chrome (web)           • chrome                               •
      web-javascript • Google Chrome 89.0.4389.90

• No issues found!
```
	
</details>

**Comment by xster** [comment]
@mkustermann could you take a look at this one? Seems like a crash in the new lightweight isolate path. I couldn't get the bot to symbolicate but the version data should be there.

**Comment by sroddy** [comment]
@xster @mkustermann I don't want to sound pessimistic but I've made a rough estimate and I calculated that this bug alone, considering an app that has on an average run about 5 to 10 FlutterViewController dismissal, would account for about 2% to 10% of the global crash rate of that given app (assuming that every possible mitigation path has been taken by the dev, otherwise it would be pretty easy to even deterministically crash almost 100% of the times).
If you are not up to trying to solve this issue ASAP (and cherry-pick to stable), I would strongly suggest to fix the docs and advise against using lightweight engines until a version that fixes this bug is released.

**Comment by mkustermann** [comment]
Was OOO during easter. Will take a look at this.

**Comment by mkustermann** [comment]
Have taken a look at it. The machine code causing the segfault looks like this:
```
dart::Isolate::ScheduleThread(this, bool, bool)
  000000000051ce3c	a9ba6ffc	stp	x28, x27, [sp, #-0x60]!
  000000000051ce40	a90167fa	stp	x26, x25, [sp, #0x10]
  000000000051ce44	a9025ff8	stp	x24, x23, [sp, #0x20]
  000000000051ce48	a90357f6	stp	x22, x21, [sp, #0x30]
  000000000051ce4c	a9044ff4	stp	x20, x19, [sp, #0x40]
  000000000051ce50	a9057bfd	stp	x29, x30, [sp, #0x50]
  000000000051ce54	910143fd	add	x29, sp, #0x50
  000000000051ce58	d11103ff	sub	sp, sp, #0x440
  000000000051ce5c	aa0203f4	mov	x20, x2
  000000000051ce60	aa0103f5	mov	x21, x1
  000000000051ce64	aa0003f3	mov	x19, x0             ### x19 = isolate
  000000000051ce68	90000f48	adrp	x8, 488 ; 0x704000
  000000000051ce6c	f9418508	ldr	x8, [x8, #0x308]
  000000000051ce70	f9400108	ldr	x8, [x8]
  000000000051ce74	f81a83a8	stur	x8, [x29, #-0x58]
  000000000051ce78	34000801	cbz	w1, 0x51cf78
  000000000051ce7c	f9403678	ldr	x24, [x19, #0x68]   ### x24 = isolate->isolate_group_
  000000000051ce80	f9400668	ldr	x8, [x19, #0x8]     ###  x8 = isolate->mutator_thread_
  000000000051ce84	b4000428	cbz	x8, 0x51cf08
  000000000051ce88	f9404908	ldr	x8, [x8, #0x90]     ###  x8 = mutator_thread->top_exit_frame_info_
                                        ^^^ Causes SEGV due to trying to load from address 0x123
```

The crashing C code is [runtime/vm/isolate.cc](https://github.com/dart-lang/sdk/blob/2.12.2/runtime/vm/isolate.cc#L677):
```c++
void IsolateGroup::IncreaseMutatorCount(Isolate* mutator) {
  ASSERT(mutator->group() == this);

  // If the mutator was temporarily blocked on a worker thread, we have to
  // unblock the worker thread again.
  Thread* mutator_thread = mutator->mutator_thread();
  if (mutator_thread != nullptr && mutator_thread->top_exit_frame_info() != 0) {
    thread_pool()->MarkCurrentWorkerAsUnBlocked();
  }
```
The `isolate->mutator_thread_` pointer has an invalid address (`0x93`).

This indicates that the `isolate` pointer passed by the flutter engine to `Dart_EnterIsolate` is invalid - because the `isolate->mutator_thread_` is either `NULL` or it points to a valid `Thread` object until the isolate dies.

(Since the stable branch this code has been modified a little in [](https://dart-review.googlesource.com/c/sdk/+/186287) - but that is not really relevant to this bug I believe.)

Now looking at the stack more closely, we can see this gets indirectly called by a lambda that calls `tonic::DartPersistentValue::Clear()`.
```
#7	0x0000000105a76f6c in tonic::DartPersistentValue::Clear() at /opt/s/w/ir/cache/builder/src/flutter/third_party/tonic/dart_persistent_value.cc:49
#8	0x0000000105a01d94 in std::__1::__function::__value_func<void ()>::operator()() const [inlined] at /opt/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/functional:1860
#9	0x0000000105a01d88 in std::__1::function<void ()>::operator()() const [inlined] at /opt/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/functional:2419
#10	0x0000000105a01d88 in fml::MessageLoopImpl::FlushTasks(fml::FlushType) at /opt/s/w/ir/cache/builder/src/flutter/fml/message_loop_impl.cc:130
#11	0x0000000105a03bf4 in fml::MessageLoopImpl::RunExpiredTasksNow() [inlined] at /opt/s/w/ir/cache/builder/src/flutter/fml/message_loop_impl.cc:143
#12	0x0000000105a03bec in fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) at /opt/s/w/ir/cache/builder/src/flutter/fml/platform/darwin/message_loop_darwin.mm:75
#13	0x000000018044dfa0 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ ()
#14	0x000000018044dba0 in __CFRunLoopDoTimer ()
#15	0x000000018044cffc in __CFRunLoopDoTimers ()
#16	0x0000000180446ee4 in __CFRunLoopRun ()
```

What I suspect is that the engine doesn't correctly handle the isolate's lifecycle: Once an isolate has died its resources are freed and one cannot use the `Dart_Isolate` anymore. The implementation of `Clear()` looks like this:

```c++
void DartPersistentValue::Clear() {
  if (!value_) {
    return;
  }

  auto dart_state = dart_state_.lock();
  if (!dart_state) {
    return;
  }

  if (Dart_CurrentIsolateGroup()) {
    Dart_DeletePersistentHandle(value_);
  } else {
    DartIsolateScope scope(dart_state->isolate());
    Dart_DeletePersistentHandle(value_);
  }
  dart_state_.reset();
  value_ = nullptr;
}
```

So it will enter an isolate and then delete the weak persistent handle. My suspicion is that the `dart_state->isolate()` has already died by the time this lambda gets executed, making this effectively trigger a use-after-free in the VM.

There's two places where this is called from afaik:

  * [lib/ui/painting/multi_frame_codec.cc](https://github.com/flutter/engine/blob/56b13559807bf4cb83d16c659757d6ae4f209490/lib/ui/painting/multi_frame_codec.cc#L182)
  * [lib/ui/window/platform_message_response_dart.cc](https://github.com/flutter/engine/blob/56b13559807bf4cb83d16c659757d6ae4f209490/lib/ui/window/platform_message_response_dart.cc#L26)

Due to lack of deep flutter engine implementation details, I don't know how the isolate lifecycle is managed.

@xster @gaaclarke Do you think the findings above make sense?  Could this be caused by incorrect isolate lifecycle in the engine?

**Comment by mkustermann** [comment]
Maybe as a side note: I've tried to reproduce this on the reported version on my Android/iOS devices and simulators, but so far not succeeded. The issue is very likely to be timing dependent.

**Comment by sroddy** [comment]
Hi @mkustermann, make sure you are running either in profile or release mode and make sure that you have pushed at least 2 ViewControllers that use a single engine group before popping them (i.e. if you pop the first flutterviewcontroller it doesn't work, you need to push three times the `next` button).
If you follow my reproduction steps on a real iOS device (you cannot run in profile/release mode on simulators) you will hit it >90% of the times.

I also tried to debug this a bit and I've noticed that when using a single engine the `dart_state_.lock()` call always returns a something falsy (I think a nullptr), so that the `if (!dart_state)` on the next line always returns before executing other code. 

On the other hand, when using lightweight engines the `dart_state_.lock()` call returns an object which is in a `Shutdown` phase but still exists, thus making the `if (!dart_state)` check pass

**Comment by mkustermann** [comment]
@sroddy Ah, yes. I tried release on my Android but when switching to iOS I forgot to change the setting to release. It's easy to reproduce!

**Comment by TahaTesser** [comment]
@mkustermann 
I was just working a few min overtime after work hours to provide you all the exact steps from the beginning, now you can reproduce it, I can log off for today 😄

**Comment by gaaclarke** [comment]
@mkustermann Thanks for the thorough investigation, that sounds reasonable.  I'll investigate further.

**Comment by gaaclarke** [comment]
I've verified that:
* `dart::Isolate::mutator_thread()` is getting called on the `dart::Isolate` after `dart::Isolate::~Isolate()` has been called.
* https://github.com/flutter/engine/blob/56b13559807bf4cb83d16c659757d6ae4f209490/lib/ui/window/platform_message_response_dart.cc#L26  is where the crash is happening.

It seems that RuntimeController is [destroying the isolate](https://github.com/gaaclarke/engine/blob/e0d515369fac81e34c2d48f39c6ca766096add54/runtime/runtime_controller.cc#L102) but there are still some outstanding instances of PlatformMessageResponseDart.  When ~PlatformMessageResponseDart is called it tries to [clear out its value](https://github.com/flutter/engine/blob/56b13559807bf4cb83d16c659757d6ae4f209490/lib/ui/window/platform_message_response_dart.cc#L26), which involves [entering the stored isolate on the DartPersistentValue](https://github.com/gaaclarke/engine/blob/e0d515369fac81e34c2d48f39c6ca766096add54/third_party/tonic/dart_persistent_value.cc#L49:L49), but of course that isolate has already been shut down.

This isn't a problem caused by the FlutterEngineGroup and should have always been present, we are just seeing it now from the new usage.

**Comment by gaaclarke** [comment]
@sroddy was correct that if you execute `p ((flutter::DartIsolate *)dart_state.__ptr_)->GetPhase()` in the following frame of the crash, it will report `Shutdown`.  Unfortunately the phase isn't within the purview of tonic::DartState, which is what we have here:
```c++
void DartPersistentValue::Clear() {
  if (!value_) {
    return;
  }

  auto dart_state = dart_state_.lock();
  if (!dart_state) {
    return;
  }

  if (Dart_CurrentIsolateGroup()) {
    Dart_DeletePersistentHandle(value_);
  } else {
    printf("phase: %d\n", ((flutter::DartIsolate*)dart_state.get())->GetPhase());
    DartIsolateScope scope(dart_state->isolate());
    Dart_DeletePersistentHandle(value_);
  }
  dart_state_.reset();
  value_ = nullptr;
}
```

I drew up this image previously that shows the relationship between everything the Flutter.  The PlatformMessageResponseDart is holding onto a `DartState`, the phase information only lives on the `DartIsolate`:
![dart_flutter](https://user-images.githubusercontent.com/30870216/113937772-8405ac00-97ae-11eb-939b-831f2df6362c.png)

**Comment by sroddy** [comment]
I think I've also found a `IsShuttingDown()` boolean on that `dart_state` in tonic that, for some reasons, when the crash occours  is still `true` even if the phase has already been mutated.

**Comment by gaaclarke** [comment]
@sroddy Yea, i saw that too.  That looks like it's suppose to be blocking this, but it isn't.  I'm looking into it.  I'll have to be careful when editing this to not break other platforms =T

**Comment by gaaclarke** [comment]
@mkustermann I don't think we are getting shutdown callbacks for isolates in isolate groups.

Here is where we specify the callback: https://github.com/gaaclarke/engine/blob/e0d515369fac81e34c2d48f39c6ca766096add54/runtime/dart_vm.cc#L413

Here is where we call shutdown:
https://github.com/gaaclarke/engine/blob/e0d515369fac81e34c2d48f39c6ca766096add54/runtime/dart_isolate.cc#L565

I can verify that `Dart_ShutdownIsolate()` is gettin called on the isolate that has the problem, but `DartIsolate::DartIsolateShutdownCallback` isn't getting called.

`DartIsolate::DartIsolateShutdownCallback` is the one responsible for setting [`IsShuttingDown()`](https://github.com/flutter/engine/blob/f41cf17dbf1f09c0a45f0afb42b25745d8ebcdbf/runtime/dart_isolate.cc#L1122) on the DartState.  When that is set properly we can make sure we don't try to clean up values on isolates that have already been shutdown.  

The DartWeakPersistentValue appears to already be doing that, not the `DartPersistentValue` for some reason:
```c++
void DartWeakPersistentValue::Clear() {
  if (!handle_) {
    return;
  }

  auto dart_state = dart_state_.lock();
  if (!dart_state) {
    // The DartVM that the handle used to belong to has been shut down and that
    // handle has already been deleted.
    handle_ = nullptr;
    return;
  }

  // The DartVM frees the handles during shutdown and calls the finalizers.
  // Freeing handles during shutdown would fail.
  if (!dart_state->IsShuttingDown()) {
    if (Dart_CurrentIsolateGroup()) {
      Dart_DeleteWeakPersistentHandle(handle_);
    } else {
      // If we are not on the mutator thread, this will fail. The caller must
      // ensure to be on the mutator thread.
      DartIsolateScope scope(dart_state->isolate());
      Dart_DeleteWeakPersistentHandle(handle_);
    }
  }
  // If it's shutting down, the handle will be deleted already.

  dart_state_.reset();
  handle_ = nullptr;
}
```

**Comment by mkustermann** [comment]
> @mkustermann I don't think we are getting shutdown callbacks for isolates in isolate groups.

That might be due to this code in the engine [runtime/dart_isolate.cc](https://github.com/flutter/engine/blob/master/runtime/dart_isolate.cc#L267):
```c++
      return Dart_CreateIsolateInGroup(
          /*group_member=*/spawning_isolate->isolate(),
          /*name=*/(*isolate_group_data)->GetAdvisoryScriptEntrypoint().c_str(),
          /*shutdown_callback=*/nullptr,
          /*cleanup_callback=*/nullptr,
          /*child_isolate_data=*/isolate_data,
          /*error=*/error);
```
which explicitly passes `nullptr` for those callbacks. (This new API avoids using `Dart_InitializeParams` callbacks which are global to the Dart VM in the process (which was a design mistake). Instead it allows having different callbacks for different isolates and allows therefore more fine-grained control.)

> The DartWeakPersistentValue appears to already be doing that, not the DartPersistentValue for some reason:

We have to be careful here, [third_party/tonic/dart_weak_persistent_value.cc](https://github.com/flutter/engine/blob/master/third_party/tonic/dart_weak_persistent_value.cc#L29):
```c++
  // The DartVM frees the handles during shutdown and calls the finalizers.
  // Freeing handles during shutdown would fail.
  if (!dart_state->IsShuttingDown()) {
    if (Dart_CurrentIsolateGroup()) {
      Dart_DeleteWeakPersistentHandle(handle_);
    } else {
      // If we are not on the mutator thread, this will fail. The caller must
      // ensure to be on the mutator thread.
      DartIsolateScope scope(dart_state->isolate());
      Dart_DeleteWeakPersistentHandle(handle_);
    }
  }
  // If it's shutting down, the handle will be deleted already.
```

The comment says that the handles will be freed by DartVM during shutdown. That is true for the shutdown of the isolate group - which happens implicitly when the last isolate within a group dies (which destroys the isolate group's managed heap, persistent handles, ...) - but not on shutdown of an individual isolate.

Those persistent handles are not associated with an isolate, but rather with the isolate group (which owns the heap).

There's probably two paths here:

* If flutter engine keeps a "root" isolate from which new engine's isolates will be created via `Dart_CreateIsolateInGroup`, then the deletion of persistent handles can use this root isolate (if it knows the root isolate is still alive)
* We could change Dart VM to associate those persistent handles with individual isolates - though we'd need to think more about whether this is a good idea.

**Comment by gaaclarke** [comment]
Yea, it seems like we need some sort of notification to existing DartPersistentValue's that the isolate is being shutdown.  Since the memory is associated with the IsolateGroup we can't just ignore calling `Dart_DeletePersistentHandle` like we previously did.  That would cause a leak until the last isolate is shutdown.  Once the isolate is deleted there is no way to safely remove that memory.

I'm not necessarily advocating for it, but if we had an API that allowed us to bind the IsolateGroup like so, we could wouldn't have to worry about deleting the memory after the isolate is gone:
```c++
if (Dart_EnterDartIsolateGroup(isolate_group_)){
  Dart_DeletePersistentHandle(value_);
  Dart_ExitDartIsolateGroup();
}
```

**Comment by mkustermann** [comment]
> I'm not necessarily advocating for it, but if we had an API that allowed us to bind the IsolateGroup like so, we could wouldn't 
> have to worry about deleting the memory after the isolate is gone:
> ```
>   if (Dart_EnterDartIsolateGroup(isolate_group_)){
>      Dart_DeletePersistentHandle(value_);
>      Dart_ExitDartIsolateGroup();
>   }
> ```

It does seem appealing, I understand.

Though there are of course a reasons why we haven't done that. It would put the responsibility of managing the isolate group lifecycle to every embedder, which is a complex task. Embedders also don't manage isolates spawned from Dart's`Isolate.spawn()` API. It would also allow having an isolate group (possibly consuming many resources) without having any isolates in it, ...

Instead, we let the embedder use our existing API and thereby require it to "proof" the isolate group is still alive (the "proof" being an entered isolate). The good thing is that isolates are becoming very lightweight, so an embedder can always have an otherwise unused "root" isolate to do this.

**Comment by gaaclarke** [comment]
Another option for solving this is to drain and pause the ui task runner before shutting down an isolate.  The risk there is that the ui task runner is shared across multiple engines now so it could create a significant delay when shutting down the isolate.

**Comment by gaaclarke** [comment]
I've added a draft that removes the crash but leaves the sloppy memory management.  I've put some thought into fixing the memory management and it's going to be a big effort that may face some big hurdles since it will may affect the runtime performance of all Flutter usage.  I think we should fix the crash and open another issue describing the memory management issue.

**Comment by sroddy** [comment]
@gaaclarke thank you so much for the effort in fixing this!
I'm soon going to measure the potential memory leaks of this approach but, given that we are talking about a very small amount of pending platform message callbacks that might race during shutdown, the size of the leaks should be, I would say, barely noticeable even with this imperfect approach.

**Comment by mkustermann** [comment]
> Another option for solving this is to drain and block the ui task runner before shutting down an isolate. The risk there is that the 
> ui task runner is shared across multiple engines now so it could create a significant delay when shutting down the isolate.

That would be the ideal - avoiding "dangling" callbacks by ensuring all outstanding work is done before shutting down.
